h2o-core/src/main/java/water/FrameSizeMonitor.java

```diff
@@ -0,0 +1,211 @@
+package water;
+
+import water.fvec.Frame;
+import water.fvec.NewChunk;
+import water.parser.FVecParseWriter;
+import water.util.Log;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+
+public class FrameSizeMonitor extends MrFun<FrameSizeMonitor> {
+
+    private static final String LOG_LEVEL_PROP = "util.frameSizeMonitor.logLevel";
+    private static final String ENABLED_PROP = "util.frameSizeMonitor.enabled";
+    private static final String SAFE_COEF_PROP = "util.frameSizeMonitor.safetyCoefficient";
+    private static final String SAFE_FREE_MEM_DEFAULT_COEF = "0.2";
+
+    private static final int LOG_LEVEL;
+    private static final boolean ENABLED;
+    private static final float SAFE_FREE_MEM_COEF;
+    private static final int SLEEP_MS = 100;
+    private static final int MB = 1024 * 1024;
+    
+    private static final Map<Key<Job>, FrameSizeMonitor> registry = new HashMap<>();
```

Comment:
It's not serialized anywhere (currently), but I'd rather mark it as transient. To make the message clear. Otherwise It'd have to be a different implementation anyway.